### PR TITLE
ntfy-sh: build node dependencies with nodejs_18

### PR DESCRIPTION
**Pull request by arjan-s**
###### Description of changes
Build NodeJS dependencies with NodeJS 18 instead of 14, which is EOL (see https://github.com/NixOS/nixpkgs/pull/229910).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/ntfy-sh/default.nix [hunk 1]
```diff
@@ -6,7 +6,9 @@ let
   nodeDependencies = (import ./node-composition.nix {
     inherit pkgs nodejs;
     inherit (stdenv.hostPlatform) system;
-  }).nodeDependencies;
+  }).nodeDependencies.override {
+    buildInputs = with pkgs; [ pkgconfig pixman cairo pango ];
```
Comment (teutat3s) [inline pkgs/tools/misc/ntfy-sh/default.nix:10]: `pkgs.pkgconfig` does not exist. I wonder why this does not fail in CI?

```suggestion
    buildInputs = with pkgs; [ pkg-config pixman cairo pango ];
```

**Review by teutat3s — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/ntfy-sh/default.nix (hunk 1 above)
Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix:10]: `pkgconfig` does exist, and `pkg-config` is just an alias for it

Comment (teutat3s) [reply pkgs/tools/misc/ntfy-sh/default.nix:10]: `nixpkgs-review pr 230188` output:
```
error: undefined variable 'pkgconfig'

       at /home/teutat3s/.cache/nixpkgs-review/pr-230188/nixpkgs/pkgs/tools/misc/ntfy-sh/default.nix:10:32:

            9|   }).nodeDependencies.override {
           10|     buildInputs = with pkgs; [ pkgconfig pixman cairo pango ];
             |                                ^
           11|   };
(use '--show-trace' to show detailed location information)
```

Comment (lilyinstarlight) [reply pkgs/tools/misc/ntfy-sh/default.nix:10]: > `pkgconfig` does exist, and `pkg-config` is just an alias for it

It's actually the opposite. The OfBorg `ofborg-eval-package-list-no-aliases` check is supposed to be catching these but it only seems to work when the alias is a package arg, not when you try to fish it out of `pkgs`

Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix:10]: That's really weird. I tested the build successfully before submitting, and also checked the packages and alias. I'll try again tonight and update the PR if necessary.

Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix]: Okay, I figured it out and you're right. I tried to do the right thing, but in doing so I misread the aliases file. It's fixed now, and I also removed the use of `pkgs`.

**Review by teutat3s — APPROVED** [review]

**Comment by teutat3s** [comment]
Result of `nixpkgs-review pr 230188` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ntfy-sh</li>
  </ul>
</details>

File: pkgs/tools/misc/ntfy-sh/default.nix [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
+, pkg-config, pixman, cairo, pango }:
 
 
 let
   nodeDependencies = (import ./node-composition.nix {
     inherit pkgs nodejs;
     inherit (stdenv.hostPlatform) system;
-  }).nodeDependencies;
+  }).nodeDependencies.override {
+    buildInputs = [ pkg-config pixman cairo pango ];
```
Comment (lilyinstarlight) [inline pkgs/tools/misc/ntfy-sh/default.nix:11]: `pkg-config` should be a `nativeBuildInput`, right?

Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix]: Yes, you're absolutely right, thanks for catching it! I fixed it now.

File: pkgs/tools/misc/ntfy-sh/default.nix [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
 
 let
   nodeDependencies = (import ./node-composition.nix {
     inherit pkgs nodejs;
     inherit (stdenv.hostPlatform) system;
-  }).nodeDependencies;
+  }).nodeDependencies.override {
+    nativeBuildInputs = [ pkg-config ];
+    buildInputs = [ pixman cairo pango ];
+  };
```
Comment (lilyinstarlight) [inline pkgs/tools/misc/ntfy-sh/default.nix:13]: Just curious, why were these necessary? It seems to be building without these inputs already on `nodejs_18`, unless I'm misunderstanding

This looks merge-ready to me otherwise though. I was just wondering about that first

Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix:13]: Actually it doesn't build without. With `nodejs_14` it built just fine, but with 16 and 18 this was necessary. If you can get it to build without this, that would obviously be better!

Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix:13]: Huh, that's weird... I see those same results (on `x86_64-linux` nixos-unstable) from `nix why-depends`. Still, simply changing the node2nix stuff to `nodejs_18` results in build errors that can only be fixed by including the changes in `default.nix`. Perhaps because it's a sub-derivation (`node-dependencies-ntfy`) that gets built? But that also depends on `nodejs_18` in the end... I have no clue... :(

Comment (lilyinstarlight) [reply pkgs/tools/misc/ntfy-sh/default.nix:13]: Tbh I'm happy to merge as-is if it fixes it for you. It was just weird, because I did remove those lines from this branch to bw sure it still built fine, and it did (in fact it actually just substituted from the cache since without those lines this is a zero-rebuild PR)

Comment (arjan-s) [reply pkgs/tools/misc/ntfy-sh/default.nix]: I'm completely baffled. I removed these lines and like you said it built just fine. On my other system I spent at least one or two hours to figure out how to make it build again, and now it just works... :S
(sorry for the commits spam, I rebased to master before testing again)

**Review by lilyinstarlight — APPROVED** [review]
